packages/dds/tree/src/shared-tree-core/sharedTreeCore.ts

```diff
@@ -115,6 +118,7 @@ export class SharedTreeCore<TEditor extends ChangeFamilyEditor, TChange> extends
 		attributes: IChannelAttributes,
 		telemetryContextPrefix: string,
 		schemaAndPolicy: SchemaAndPolicy,
+		enricher?: CommitEnricher<TChange>,
```

Comment:
I must say I'm not a fan of this language feature\*. I usually bow down to consistency pressure when the rest of the members already do it, but that's not the case here so I didn't want to start it.

\* My main beef is that it makes the question "what are the members of this class" easier to get wrong by adding an extra place where one should go look (I regularly get tricked by this).
Another thing that I don't like about it is that it mixes public API choices with implementation detail choices. None of this is worth the terseness in my view but I can see how that might be a matter of taste.

Reply:
Well, I don't think we have any style guides around it so you can do as you please!

I personally like the pattern, but exclusively as sugar. I would never take a field that is _not_ currently passed into the constructor and make it so, just to use this language feature - that sounds like an antipattern. But if it's an input already, it just reduces the boilerplate. To your Main Beef, it is indeed true that if you remove two-thirds of the boilerplate, the remaining boilerplate becomes three times harder to find... I can't argue against that logic ;) But really, I do sympathize with that a bit. I think the best-looking classes are the ones that have _all_ their fields declared in a row (usually just above the constructor) or _all_ their fields inlined into the constructor parameters. It's so easy then to see what the state is in its entirety. But nonetheless I will always take the terser option if it's available, because it removes two lines of code per field, _and_ it allows you to use the same name for the constructor parameter and the field. That's the Main Beef I have with doing it the ol' fashioned way - you have to pick two different names for the field, one which is exposed to the consumer and the other which is exposed to the implementer. So you either end up with underscores or you have to choose slightly different names that mean the same thing.

Reply:
> I would never take a field that is not currently passed into the constructor and make it so
Oh dear, it haddn't occured to me that someone would ever do this. I can't say I'm really worried about it. Perhaps I should clarify what I mean by "mixes public API choices with implementation detail choices": you may want your ctor to take `policy?: Foo` while having a member field `policy: Foo | Bar`, or you may want your ctor to be `policy: Foo | undefined` while your member field is `policy?: Foo`. Using the language feature in question promotes a habit where the question "how should the ctor params differ from the member fields?" has a smaller chance of being given its due. Even when it's used when there is no such difference, its usage increases the chance that a future refactor of the params will end up affecting both sides (inner and outer) when it really should affect only one of them.

> you have to pick two different names for the field
Waaat? See `this.schemaAndPolicy = schemaAndPolicy`.


> I think the best-looking classes are the ones that have all their fields declared in a row or all their fields inlined into the constructor parameters.
Despite my reservations about the latter, I feel the same way and will bow down to consistency when that's the case.

Reply:
Yea forget what I said, I was saying complete nonsense. I was thinking of a scenario involving getters which was on my mind, but is completely unrelated to this. Sorry about that :D

That's a really good point that about the type and optionality of the property being different between the state and the parameter sometimes. I see it as just one more case when using this language feature simply isn't appropriate.

Reply:
Your arguments have swayed me into believing the following:
1. Using this language pattern is fine if it is "purely sugar" for what would otherwise be there.
2. But there shouldn't be any pressure to use it if you like the old fashioned way for whatever reason

Reply:
Does that mean you don't buy my maintainability argument about how it may be pure sugar at some point in time, but it'll trick you into forgetting to think about cases where it ought to stop being pure sugar when making a change? Or do you just give the convenience of the feature greater weight than this somewhat remote risk?

Reply:
I don't think I buy that argument, if I'm understanding it correctly. For example, if I see a sugared constructor parameter and I change it to be optional, I'll make that change and then see that I have a bunch of red squiggles everywhere that I read it in the class. I would then think, "ah, I should move this into a non-sugared field". I don't think I'd ever say, "well, it's a sugared constructor parameter and I don't want to change that, but I want it to be optional... I guess I'll just need to assert that it is not undefined everywhere that I read it." You can always turn a sugared parameter into a non-sugared field/parameter and visa versa without changing the class constructor API, so it seems harmless enough. Maybe you have some more subtle cases where it tricks people more dangerously, but I haven't thought of any.

Reply:
I was just thinking abstractly. I think there are cases were it could be more dangerous (e.g., where slient type cohersion will not warn you that you're now treating undefined as a boolean) but I think you're right that in most cases *some* part of the code would lead to compiler to complain. This would be a very different story in plain JS of course.